create Oscar groups with context

ThomasBreuer opened the pull request:
(This addresses issue #894.)

Several types of Oscar groups need context information besides the underlying GAP object:
`PermGroup` needs the degree, `MatrixGroup` needs the dimension and the ring of the matrix entries.
This context was not used in certain situations.

Now the recommended way to construct groups with context is via `oscar_group(gapobj, G)` where `gapobj` is a GAP group and `G` is an Oscar group that provides the context.
(Up to now we had `T(gapobj)`, where `T` is the type of `G`, but this does not work in general. It would be tempting to use `T(gapobj, G)`, but this syntax cannot be used in general because `AutomorphismGroup` uses it for something different.)

(More methods should get restricted to situations where the contexts of the arguments are equal. This will be the next step.)

ThomasBreuer committed:
create Oscar groups with context

(This addresses issue #894.)

Several types of Oscar groups need context information besides the underlying GAP object:
`PermGroup` needs the degree, `MatrixGroup` needs the dimension and the ring of the matrix entries.
This context was not used in certain situations.

Now the recommended way to construct groups with context is via `oscar_group(gapobj, G)` where `gapobj` is a GAP group and `G` is an Oscar group that provides the context.
(Up to now we had `T(gapobj)`, where `T` is the type of `G`, but this does not work in general. It would be tempting to use `T(gapobj, G)`, but this syntax cannot be used in general because `AutomorphismGroup` uses it for something different.)

(More methods should get restricted to situations where the contexts of the arguments are equal. This will be the next step.)

fieker commented:
Still not perfect...or it is user error. Failing `example`:
```
Qx, x = QQ["x"]
galois_group((x-1)*(x-2)*(x-3))
```
error:
```
ERROR: G and x are not compatible
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] conjugate_group(G::PermGroup, x::PermGroupElem)
   @ Oscar ~/.julia/dev/Oscar/src/Groups/GAPGroups.jl:543
 [3] ^(H::PermGroup, y::PermGroupElem)
   @ Oscar ~/.julia/dev/Oscar/src/Groups/GAPGroups.jl:547
 [4] galois_group(f::fmpq_poly; pStart::Int64)
```

ThomasBreuer committed:
changed the degree of `inner_direct_product`

ThomasBreuer commented:
@fieker commit 0ffc2a6 fixes the handling of examples as shown above, in the sense that the counting of points now fits.

ThomasBreuer committed:
forbid `PermGroup(G, 0)`

We do not support `symmetric_group(0)`,
which means that the degree of a permutation group must be positive.

fieker commented:
On Thu, Feb 17, 2022 at 03:13:35AM -0800, Thomas Breuer wrote:
> @fieker commit 0ffc2a6 fixes the handling of examples as shown above, in the sense that the counting of points now fits.
> 
It works! Thanks
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/1094#issuecomment-1042839551
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>
